Library/Homebrew/utils/curl.rb

```diff
@@ -213,18 +213,18 @@ def curl_headers(*args, wanted_headers: [], **options)
         )
 
         # 22 means a non-successful HTTP status code, not a `curl` error, so we still got some headers.
-        if result.success? || result.exit_status == 22
-          parsed_output = parse_curl_output(result.stdout)
+        next if !result.success? && result.exit_status != 22
 
-          # If we didn't get any wanted header yet, retry using `GET`.
-          next if request_args.empty? && wanted_headers.any? &&
-                  parsed_output.fetch(:responses).none? { |r| (r.fetch(:headers).keys & wanted_headers).any? }
+        parsed_output = parse_curl_output(result.stdout)
 
-          return parsed_output if result.success?
-        end
+        # If we didn't get any wanted header yet, retry using `GET`.
+        next if request_args.empty? && wanted_headers.any? &&
+                parsed_output.fetch(:responses).none? { |r| (r.fetch(:headers).keys & wanted_headers).any? }
 
-        result.assert_success!
+        return parsed_output if result.success?
```

Comment:
This changes the return type of this function to be nilable instead of raising an error.

Reply:
It should still raise in download strategies.